src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al

```diff
@@ -0,0 +1,604 @@
+    local procedure EnsureValidLanguage(Language: Text): Text
+    var
+        NotSupportedNotification: Notification;
+    begin
+        case LowerCase(Language) of
+            'en', 'cs', 'de', 'es', 'fr', 'hu', 'it', 'ja', 'ko', 'nl', 'pl', 'pt-br', 'pt-pt', 'ru', 'sv', 'tr', 'zh-hans', 'zh-hant':
+                exit(LowerCase(Language));
+            else begin
+                NotSupportedNotification.Id := '0664870f-bd05-46cc-9e98-cc338d7fdc64';
+                NotSupportedNotification.Message(StrSubstNo(UnsupportedLanguageNotificationLbl, Language));
+                NotSupportedNotification.Send();
+
+                exit('en');
+            end;
+        end;
+    end;
+
+    #endregion
+
+    local procedure InsertProductFromObject(Offer: JsonObject; var Product: Record "AppSource Product")
+    begin
+        Product.Init();
+        Product.UniqueProductID := CopyStr(GetStringValue(Offer, 'uniqueProductId'), 1, MaxStrLen(Product.UniqueProductID));
+        Product.DisplayName := CopyStr(GetStringValue(Offer, 'displayName'), 1, MaxStrLen(Product.DisplayName));
+        Product.PublisherID := CopyStr(GetStringValue(Offer, 'publisherId'), 1, MaxStrLen(Product.PublisherID));
+        Product.PublisherDisplayName := CopyStr(GetStringValue(Offer, 'publisherDisplayName'), 1, MaxStrLen(Product.PublisherDisplayName));
+        Product.PublisherType := CopyStr(GetStringValue(Offer, 'publisherType'), 1, MaxStrLen(Product.PublisherType));
+        Product.RatingAverage := GetDecimalValue(Offer, 'ratingAverage');
+        Product.RatingCount := GetIntegerValue(Offer, 'ratingCount');
+        Product.ProductType := CopyStr(GetStringValue(Offer, 'productType'), 1, MaxStrLen(Product.ProductType));
+        Product.Popularity := GetDecimalValue(Offer, 'popularity');
+        Product.LastModifiedDateTime := GetDateTimeValue(Offer, 'lastModifiedDateTime');
+
+        Product.AppID := ExtractAppIDFromUniqueProductID(Product.UniqueProductID);
+
+        // Insert, if it fails to insert due to the data (ex duplicate ids), ignore the error
+        if not Product.Insert() then;
+    end;
+
+    [NonDebuggable]
+    local procedure GetAPIKey(): SecretText
+    var
+        ApiKey: SecretText;
+    begin
+        Init();
+        if not Dependencies.EnvironmentInformation_IsSaas() then
+            Error(NotSupportedOnPremisesErrorLbl);
+
+        Dependencies.AzureKeyVault_GetAzureKeyVaultSecret('MS-AppSource-ApiKey', ApiKey);
+        exit(ApiKey);
+    end;
+
+    local procedure Init()
+    begin
+        if not IsDependenciesSet then
+            SetDefaultDependencyImplementation();
+    end;
+
+    local procedure SetDefaultDependencyImplementation()
+    var
+        dependencyInstance: Codeunit "AppSource Product Manager";
+    begin
+        dependencies := dependencyInstance;
+        IsDependenciesSet := true;
+    end;
+
+    var
```

Comment:
Put the global parameters at the top of the file before functions.

Reply:
Next PR